warehouse/query-core/src/main/java/datawave/query/jexl/JexlNodeBuilder.java

```diff
@@ -0,0 +1,637 @@
+    }
+    
+    /**
+     * Set the value for the node this builder will create.
+     * 
+     * @param value
+     *            the value
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withValue(Object value) {
+        this.value = value;
+        return this;
+    }
+    
+    /**
+     * Add a child node to this builder.
+     * 
+     * @param node
+     *            the child node
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withChild(JexlNode node) {
+        children.add(() -> node);
+        return this;
+    }
+    
+    /**
+     * Add each node in the provided vararg as a child to this builder.
+     * 
+     * @param children
+     *            the children nodes
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withChildren(JexlNode... children) {
+        for (JexlNode child : children) {
+            withChild(child);
+        }
+        return this;
+    }
+    
+    /**
+     * Add each node in the provided iterable as a child to this builder.
+     * 
+     * @param children
+     *            the children nodes
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withChildren(Iterable<? extends JexlNode> children) {
+        children.forEach(this::withChild);
+        return this;
+    }
+    
+    /**
+     * Add a child node builder to this builder. The final child will be built when {@link JexlNodeBuilder#build()} is called on this builder.
+     * 
+     * @param builder
+     *            the builder
+     * @return this builder
+     */
+    public JexlNodeBuilder<T> withChild(JexlNodeBuilder<?> builder) {
+        children.add(builder::build);
+        return this;
+    }
+    
+    /**
+     * Build and return the new {@link JexlNode} instance with the specified image, value, and children. Any child nodes or child node suppliers that were
+     * provided to this builder that are either null or result in null nodes will not be included in the node's children.
+     * 
+     * @return the node
+     */
+    public T build() {
+        // Create the initial node.
+        T node = constructor.get();
+        node.image = this.image;
+        node.jjtSetValue(this.value);
+        
+        // Create the finalized children.
+        // @formatter:off
+        JexlNode[] childArray = children.stream()
+                        .filter(Objects::nonNull) // Filter out any possible null suppliers.
+                        .map(Supplier::get)       // Get the actual child nodes.
+                        .filter(Objects::nonNull) // Filter out any null children.
+                        .toArray(JexlNode[]::new);
+        // @formatter:on
+        
+        // Set the parent for each child.
+        for (JexlNode child : childArray) {
+            child.jjtSetParent(node);
+        }
+        // Add each child to the root node.
+        for (int i = 0; i < childArray.length; i++) {
+            node.jjtAddChild(childArray[i], i);
```

Comment:
This is a really bad call to make.  The underlying implementation creates a copy of the array with one additional entry , and copies all of the elements each time you call jjtAddChild.  A better way of adding children to a parent node is to call JexlNodes.children(T node, JexlNode... children).  If you change the code to use this then you no longer need code to set the parent for each child - that is already handled by the call.

Reply:
Good catch, I will update this to use JexlNodes.children instead.